barretenberg/cpp/src/barretenberg/vm2/simulation/update_check.test.cpp

```diff
@@ -0,0 +1,305 @@
+
+    uint32_t block_number = 100;
+    ContractInstance instance = testing::random_contract_instance();
+    instance.current_class_id = param.current_class_id;
+    instance.original_class_id = param.original_class_id;
+
+    AztecAddress derived_address = compute_contract_address(instance);
+    FF shared_mutable_slot = poseidon2::hash({ UPDATED_CLASS_IDS_SLOT, derived_address });
+    FF shared_mutable_hash_slot = shared_mutable_slot + UPDATES_SHARED_MUTABLE_VALUES_LEN;
+    FF shared_mutable_leaf_slot =
+        poseidon2::hash({ GENERATOR_INDEX__PUBLIC_LEAF_INDEX, DEPLOYER_CONTRACT_ADDRESS, shared_mutable_hash_slot });
+
+    FF update_metadata = FF(static_cast<uint64_t>(123) << 32) + param.update_block_of_change;
+    std::vector<FF> update_preimage = { update_metadata, param.update_pre_class, param.update_post_class };
+    std::vector<FF> update_preimage_slots;
+
+    for (size_t i = 0; i < update_preimage.size(); ++i) {
+        FF leaf_slot =
+            poseidon2::hash({ GENERATOR_INDEX__PUBLIC_LEAF_INDEX, DEPLOYER_CONTRACT_ADDRESS, shared_mutable_slot + i });
+        update_preimage_slots.push_back(leaf_slot);
+    }
+
+    FF update_hash = poseidon2::hash(update_preimage);
+
+    TreeSnapshots trees;
+    trees.publicDataTree.root = 42;
+
+    NiceMock<MockPoseidon2> poseidon2;
+    NiceMock<MockHighLevelMerkleDB> merkle_db;
+    NiceMock<MockLowLevelMerkleDB> mock_low_level_merkle_db;
+    NiceMock<MockRangeCheck> range_check;
+
+    EventEmitter<UpdateCheckEvent> event_emitter;
+    UpdateCheck update_check(poseidon2, range_check, merkle_db, block_number, event_emitter);
+
+    EXPECT_CALL(merkle_db, storage_read(shared_mutable_leaf_slot)).WillRepeatedly(Return(update_hash));
+    EXPECT_CALL(merkle_db, get_tree_roots()).WillRepeatedly(ReturnRef(trees));
+    EXPECT_CALL(merkle_db, as_unconstrained()).WillRepeatedly(ReturnRef(mock_low_level_merkle_db));
+
+    EXPECT_CALL(mock_low_level_merkle_db, get_low_indexed_leaf(world_state::MerkleTreeId::PUBLIC_DATA_TREE, _))
+        .WillRepeatedly([&](world_state::MerkleTreeId, const FF& leaf_slot) {
+            for (size_t i = 0; i < update_preimage_slots.size(); ++i) {
+                if (leaf_slot == update_preimage_slots[i]) {
+                    return GetLowIndexedLeafResponse(true, static_cast<uint64_t>(i));
+                }
+            }
+            throw std::runtime_error("Leaf not found");
+        });
+
+    EXPECT_CALL(mock_low_level_merkle_db, get_leaf_preimage_public_data_tree(_))
+        .WillRepeatedly([&](const uint64_t& index) {
+            return PublicDataTreeLeafPreimage(
+                PublicDataLeafValue(FF(index) + shared_mutable_leaf_slot, update_preimage[index]), 0, 0);
+        });
+
+    EXPECT_CALL(poseidon2, hash(_)).WillRepeatedly([&](const std::vector<FF>& input) {
+        return poseidon2::hash(input);
+    });
+
+    EXPECT_CALL(range_check, assert_range(_, _)).WillRepeatedly([&](const uint128_t& value, const uint8_t& range) {
```

Comment:
Hmm, we might want (not in this PR) to consider writing non-event-emmiting versions of these that we can just use here. They might just be the foundation for the standalone simulator.